### DISCUSSION TITLE
DeclarativeBaseNoMeta with abstract classes?

### DISCUSSION THREAD

**Discussion by pamelafox** [sqlalchemy/sqlalchemy#9861] [Usage Questions] [answered]
I am running some tests in Flask-SQLAlchemy with the new Declarative bases, and I ran into an issue that I can also replicate with a tweak of the documentation example:

```
from sqlalchemy.orm import DeclarativeBase, DeclarativeBaseNoMeta
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy import Integer, String


class Base(DeclarativeBaseNoMeta):
    pass


class Employee(Base):
    __abstract__ = True


class Manager(Employee):
    __tablename__ = "manager"
    id = mapped_column(Integer, primary_key=True)
    name = mapped_column(String(50))
    manager_data = mapped_column(String(40))


class Engineer(Employee):
    __tablename__ = "engineer"
    id = mapped_column(Integer, primary_key=True)
    name = mapped_column(String(50))
    engineer_info = mapped_column(String(40))
```

That code works fine for `DeclarativeBase`, but produces this error with `DeclarativeBaseNoMeta`:

```
AttributeError: type object 'Employee' has no attribute '__mapper__'
```

That error comes from map_declaratively attempting to return `cls.__mapper__`. As far as I can tell from my tracing of the code, it should be expected that the Employee class doesn't have `__mapper__`, since it's abstract, and the issue is just that the particular line of code doesn't account for that. In the `DeclarativeBase` flow, it is accounted for. The error goes away when I add a check for abstract before attempting to return `__mapper__`.

So I think this is an issue, but wanted to post as a discussion before filing.

**Comment by zzzeek** [comment, marked as answer]
> So I think this is an issue, but wanted to post as a discussion before filing.

Thanks for following our instructions on this!   at this point, I welcome you to create issues directly when you see them as such , as in this case, since yes this is two bugs, one that the code is wrong and the other that we dont have adequate test coverage for the DeclarativeBaseNoMeta class.

this is the fix for the code itself:

```diff
diff --git a/lib/sqlalchemy/orm/decl_api.py b/lib/sqlalchemy/orm/decl_api.py
index a3a817162..2301ab500 100644
--- a/lib/sqlalchemy/orm/decl_api.py
+++ b/lib/sqlalchemy/orm/decl_api.py
@@ -944,7 +944,7 @@ class DeclarativeBaseNoMeta(inspection.Inspectable[InstanceState[Any]]):
             _check_not_declarative(cls, DeclarativeBaseNoMeta)
             _setup_declarative_base(cls)
         else:
-            cls._sa_registry.map_declaratively(cls)
+            _as_declarative(cls._sa_registry, cls, cls.__dict__)
 
 
 def add_mapped_attribute(

```

**Reply by pamelafox** [reply]
Ah, I was wondering why DeclarativeBase and DeclarativeBaseNoMeta had different code paths there. I'll continue running the Flask-SQLAlchemy tests and file issues if I see anything else that's clearly an issue.
